### PR TITLE
Add torchtune docs

### PR DESCRIPTION
**Pull request by tcapelle**
Add [torchtune](https://pytorch.org/torchtune/stable/deep_dives/wandb_logging.html) documentation.

### CODE REVIEW

**Comment by tcapelle** [comment]
image
image
image

File: docs/guides/integrations/torchtune.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
... large language models (LLMs). It focuses on simplicity and extensibility with a componentized, native-PyTorch design that avoids unnecessary abstractions and promotes easy reuse. The library ensures correctness and stability, setting high standards for component testing and performance benchmarks. Torchtune is built to democratize LLM fine-tuning, offering out-of-the-box functionality across different hardware setups. It features modular implementations of popular LLMs, interoperability through checkpoint conversion, and a variety of training recipes. The library also integrates with external datasets and evaluation tools, supports distributed training, and allows configuration through YAML files to adjust training settings and hyperparameters without modifying code. Additionally, torchtune has built-in support for [logging with Weights & Biases](https://pytorch.org/torchtune/stable/deep_dives/wandb_logging.html), enhancing tracking and visualization of training processes. With a focus on usability, torchtune adheres to PyTorch’s philosophy by emphasizing clarity and modular building blocks over monolithic design, aiming to make fine-tuning accessible and efficient for developers.
```
Comment (morganmcg1) [inline docs/guides/integrations/torchtune.md]: Make this 2 sentences

File: docs/guides/integrations/torchtune.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+---
+displayed_sidebar: default
+---
+
+
+# torchtune
+
```
Comment (morganmcg1) [inline docs/guides/integrations/torchtune.md:10]: Add colab badge

File: docs/guides/integrations/torchtune.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
```
Comment (morganmcg1) [inline docs/guides/integrations/torchtune.md:13]: Add small codeblock showing the core lines of code needed to turn on W&B logging

https://docs.wandb.ai/guides/integrations/huggingface#-next-level-logging-in-few-lines

File: docs/guides/integrations/torchtune.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+:::info
+Check the [torchtune documentation](https://pytorch.org/torchtune/stable/deep_dives/wandb_logging.html) on how to use Weights & Biases to tune your LLM models.
+:::
+
+
+## Using the Weights & Biases metric logger
+
+You can quickly try the W&B integration with torchtune by overriding arguments at launch time.
+
+1. First install the `wandb` library:
```
Comment (morganmcg1) [inline docs/guides/integrations/torchtune.md]: Delete install step

File: docs/guides/integrations/torchtune.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+tune run lora_finetune_single_device --config llama3/8B_lora_single_device \
+  metric_logger._component_=torchtune.utils.metric_logging.WandBLogger \
+  metric_logger.project="llama3_lora" \
+  metric_logger.entity="my_project" \
+  metric_logger.job_type="lora_finetune_single_device" \
+  metric_logger.group="my_awesome_experiments" \
+  log_every_n_steps=5
+```
+
+## What do we log?
```
Comment (morganmcg1) [inline docs/guides/integrations/torchtune.md]: Add Table with all arguments and descriptions, like here https://docs.wandb.ai/guides/integrations/huggingface#additional-wb-settings

File: docs/guides/integrations/torchtune.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+```bash
+pip install -U wandb
+```
+
+2. Then, run any default recipe and pass the `metric_logger._component_` with the corresponding `WandBLogger` class. You can also pass a `project` name and `log_every_n_steps` to log metrics every n steps.
+
+```bash
+tune run lora_finetune_single_device --config llama3/8B_lora_single_device \
+  metric_logger._component_=torchtune.utils.metric_logging.WandBLogger \
```
Comment (morganmcg1) [inline docs/guides/integrations/torchtune.md:34]: show the part of the config file with wandb logging turned on, not just cli argument

File: docs/guides/integrations/torchtune.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+:::info
+This is a fast evolving library, the current metrics are subject to change. If you want to add a custom metric, you should modify the recipe and call the corresponding `self._metric_logger.*` function.
+:::
+
+## Saving and loading checkpoints
+
+The torchtune library supports various [checkpoint formats](https://pytorch.org/torchtune/stable/deep_dives/checkpointer.html). Depending on the origin of the model you are using, you should switch to the appropriate [checkpointer class](https://pytorch.org/torchtune/stable/deep_dives/checkpointer.html).
+
+If you want to save the model checkpoints to [W&B Artifacts](https://docs.wandb.ai/guides/artifacts), the simplest solution is to override the [`save_checkpoint`](https://github.com/pytorch/torchtune/blob/cd779783f9acecccbebc3c50265f6caf97fa99aa/recipes/full_finetune_single_device.py#L348) functions inside the corresponding recipe.
```
Comment (morganmcg1) [inline docs/guides/integrations/torchtune.md]: don't link to torchtune code as the links will probably go stale pretty quickly as its in active development

File: docs/guides/integrations/torchtune.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+```python
+def save_checkpoint(self, epoch: int) -> None:
+    ...
+    ## Let's save the checkpoint to W&B
+    ## depending on the Checkpointer Class the file will be named differently
+    ## Here is an example for the full_finetune case
+    checkpoint_file = Path.joinpath(
+        self._checkpointer._output_dir, f"torchtune_model_{epoch}"
+    ).with_suffix(".pt")
+    wandb_at = wandb.Artifact(
```
Comment (morganmcg1) [inline docs/guides/integrations/torchtune.md]: rename `wandb_at` to `art`or similr,  `*_at` isn't commonly used

**Review by morganmcg1 — CHANGES_REQUESTED** [review]

File: docs/guides/integrations/torchtune.md (hunk 8 above)
Comment (tcapelle) [reply docs/guides/integrations/torchtune.md]: I use `*_at` all the time, for some reason I always thought that this was the "standard" naming 🤔

File: docs/guides/integrations/torchtune.md (hunk 7 above)
Comment (tcapelle) [reply docs/guides/integrations/torchtune.md]: I will keep the checkpointer link, as I don't think they will change this paradigm. Will remove specific line pointers

File: docs/guides/integrations/torchtune.md (hunk 6 above)
Comment (tcapelle) [reply docs/guides/integrations/torchtune.md:34]: this makes you use W&B straight away with the supplied recipes, what do you think about adding the part of the config also?

File: docs/guides/integrations/torchtune.md (hunk 2 above)
Comment (tcapelle) [reply docs/guides/integrations/torchtune.md:10]: we don't have a colab... should we make one? this is a CLI library anyway

File: docs/guides/integrations/torchtune.md (hunk 6 above)
Comment (morganmcg1) [reply docs/guides/integrations/torchtune.md:34]: The goal is to show best practice of how to use the integration, which I think should be via a config file given that there are a number of different arguments to be set

you can show both but we should prefer the config file

File: docs/guides/integrations/torchtune.md (hunk 7 above)
Comment (morganmcg1) [reply docs/guides/integrations/torchtune.md]: sounds good

File: docs/guides/integrations/torchtune.md (hunk 8 above)
Comment (morganmcg1) [reply docs/guides/integrations/torchtune.md]: less so in integrations land I guess, 

transformers:


Screenshot 2024-04-29 at 12 06 46


lightning:


Screenshot 2024-04-29 at 12 07 24

File: docs/guides/integrations/torchtune.md (hunk 2 above)
Comment (morganmcg1) [reply docs/guides/integrations/torchtune.md:10]: yep lets create a colab

File: docs/guides/integrations/torchtune.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
+---
+displayed_sidebar: default
+---
+
+
+# torchtune
```
Comment (morganmcg1) [inline docs/guides/integrations/torchtune.md]: Can you also address the the `vale` grammar failures and warnings thanks

**Review by morganmcg1 — CHANGES_REQUESTED** [review]

File: docs/guides/integrations/torchtune.md (hunk 5 above)
Comment (tcapelle) [reply docs/guides/integrations/torchtune.md]: What do you expect here exactly? We could add a list of the params to init? as the library is evolving, for the time being we log whatever they call inside the train function, so loss, tokens_per_sec, lr and the config

File: docs/guides/integrations/torchtune.md (hunk 6 above)
Comment (tcapelle) [reply docs/guides/integrations/torchtune.md:34]: added a tabbed view

File: docs/guides/integrations/torchtune.md (hunk 7 above)
Comment (tcapelle) [reply docs/guides/integrations/torchtune.md]: roger

File: docs/guides/integrations/torchtune.md (hunk 4 above)
Comment (tcapelle) [reply docs/guides/integrations/torchtune.md]: done

**Comment by tcapelle** [comment]
New look:
image

**Comment by tcapelle** [comment]
torchtune_docs

**Comment by tcapelle** [comment]
This is ready for review/merge @ngrayluna

**Comment by ngrayluna** [comment]
Some notes:

* We don't use "->".
*  Bold text is reserved only for referring to UI elements.
* Don't use "we" in the docs.

**Review by ngrayluna — APPROVED** [review]

**Review by tcapelle — COMMENTED** [review]
the pytorch team wanted us to explicitely call torchtune with undercase.

**Comment by tcapelle** [comment]
feel free to merge @ngrayluna
